device.c

```diff
@@ -316,25 +324,42 @@ int adev_open_output_stream(struct audio_hw_device *dev,
         return -EINVAL;
     }
     /* check requested device */
-    /* TODO we should handle only devices described in XML configuration and "default" one */
     if ((devices & AUDIO_DEVICE_BIT_IN) != 0) {
         ALOGE("Failed. Incorrect device type. -EINVAL");
         *stream_out = NULL;
         pthread_mutex_unlock(&xdev->lock);
         return -EINVAL;
     }
 
-    /* TODO define card_id and dev_id for stream, using provided parameters:
-       devices, flags, address, source */
+    /* For now we have one card, so card id is always constant */
+    /* Get device id. We expect (same as parser in car audio service)
+     * that address has format "busX_%s" */
+    if (sscanf(address, "bus%u", &dev_id) != 1) {
+        ALOGE("Address has not supported format.");
+        dev_id = NUMBER_OF_DEVICES_OUT;
+    }
 
-    res = out_create(dev, handle, devices, card_id, dev_id, config, stream_out);
-    if (*stream_out != NULL) {
-        xdev->xout_stream = (x_stream_out_t*)(*stream_out);
-        if (xdev->master_is_muted) {
-            out_set_mute(xdev->xout_stream, true);
+    if (dev_id < NUMBER_OF_DEVICES_OUT) {
+        if (xdev->xout_streams[dev_id] == 0) {
+            /* create new stream on free device */
+            res = out_create(dev, handle, devices, CARD_ID, dev_id, config, stream_out);
+            if (*stream_out != NULL) {
+                xdev->xout_streams[dev_id] = (x_stream_out_t*)(*stream_out);
+                if (xdev->master_is_muted) {
+                    out_set_mute(xdev->xout_streams[dev_id], true);
+                }
+                LOG_FN_PARAMETERS("Created stream_out:%p", *stream_out);
+            }
+            /* if out_create failed then 'res' has 'failed' value and it will be reported up */
+        } else {
+            /* May be we have to return existing stream here? */
```

Comment:
Please consider adding "FIXME" tag

Reply:
This comment is removed. It was my internal note.